format/udf-spec.md

```diff
@@ -0,0 +1,301 @@
+| *required*  | `timestamp-ms`    | `long` (epoch millis)                                                                                   | Creation timestamp of this version.                            |
+
+Note:
+
+`on-null-input` provides an optimization hint for query engines, its value can be either `"returns_null_on_null_input"` or `"called_on_null_input"`:
+1. If set to `returns_null_on_null_input`, the function always returns `NULL` if any input argument is `NULL`. This allows engines to apply predicate pushdown or skip function evaluation for rows with `NULL` inputs. For a function `f(x, y) = x + y`,
+the engine can safely rewrite `WHERE f(a,b) > 0` as `WHERE a IS NOT NULL AND b IS NOT NULL AND f(a,b) > 0`.
+2. If set to `called_on_null_input`, the function may handle `NULL`s internally (e.g., `COALESCE`, `NVL`, `IFNULL`), so the engine must execute the function even if some inputs are `NULL`.
+
+### Representation
+A representation encodes how the definition version is expressed in a specific SQL dialect.
+
+| Requirement | Field name   | Type              | Description                                                                                 |
+|-------------|--------------|-------------------|---------------------------------------------------------------------------------------------|
+| *required*  | `type`       | `string`          | Must be `"sql"`                                                                             |
+| *required*  | `dialect`    | `string`          | SQL dialect identifier (e.g., `"spark"`, `"trino"`).                                        |
+| *required*  | `body`       | `string`          | SQL expression text.                                                                        |
+
+Note: The `body` must be valid SQL in the specified dialect; validation is the responsibility of the consuming engine.
+
+### Definition log
+| Requirement | Field name            | Type                                               | Description                                                      |
+|-------------|-----------------------|----------------------------------------------------|------------------------------------------------------------------|
+| *required*  | `timestamp-ms`        | `long` (epoch millis)                              | When the definition snapshot was created or updated.             |
+| *required*  | `definition-versions` | `list<{ definition-id: string, version-id: int }>` | Mapping of each definition to its selected version at this time. |
+
+## Function Resolution in Engines
+Resolution rule is decided by engines, but engines SHOULD:
+1. Prefer exact signature matches over casting or subtyping.
+2. Allow numeric widening (e.g., `INT` --> `BIGINT/FLOAT`) when needed and where it does not conflict with the rule above.
```

Comment:
Maybe I miss something, not sure I see what situation would it conflict with rule above?

Also, is this tied to Iceberg schema evolution rules, or is it separate?